### ISSUE TITLE
"TypeError: common.register is not a function" under node v20.6.0

### ISSUE THREAD

**Issue by bkuri** [shelljs/shelljs#1133]
### Operating system:

`Linux bkuri-desktop 6.5.1-AMD-znver2 #1 SMP PREEMPT_DYNAMIC Sat Sep  2 07:47:03 UTC 2023 x86_64 GNU/Linux`

### Description of the bug:

Running my cli app under node `v20.6.0` gives me the following output:

```
/home/.../node_modules/shelljs/src/cat.js:4
common.register('cat', _cat, {
       ^

TypeError: common.register is not a function
    at Object.<anonymous> (/home/.../node_modules/shelljs/src/cat.js:4:8)
    at Module._compile (node:internal/modules/cjs/loader:1241:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1295:10)
    at Module.load (node:internal/modules/cjs/loader:1091:32)
    at Module._load (node:internal/modules/cjs/loader:938:12)
    at Module.require (node:internal/modules/cjs/loader:1115:19)
    at require (node:internal/modules/helpers:130:18)
    at /home/.../node_modules/shelljs/shell.js:25:3
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/home/.../node_modules/shelljs/shell.js:24:23)

Node.js v20.6.0
error Command failed with exit code 1.
```

The error goes away after downgrading to node `v20.5.1`.

### Example Shell command to reproduce the error:

```sh
$ nvm use 20.6
$ yarn start # TypeError...
$ nvm use 20.5.1
$ yarn start
```

Hope it's not too vague of a report… Let me know if you need more details :+1:

**Comment by chase-moskal** [comment]
i can confirm this bug is a thing, i just upgraded to node 20.6 and found the same thing

affected
- https://github.com/benevolent-games/turtle/issues/16
- https://github.com/jhipster/generator-jhipster/issues/23417

will downgrade to 20.5 in the meantime

**Comment by nfischer** [comment]
Thanks for the report. We definitely want to support v20, so hopefully we can resolve this quickly. AFAIK we are exporting `common.register` the way we're supposed to, so I'm thinking this might be a nodejs bug:

https://github.com/shelljs/shelljs/blob/f7a7c75d4d315fc9fada1b148668811cf505f152/src/common.js#L496

---

I am **not** currently able to reproduce this bug. I'm using node v20.6.0 on Ubuntu 22.04. Am I missing something to reproduce?

```shell
$ node --version
v20.6.0

$ cat index.js
var shell = require('shelljs');

console.log(shell.head('~/.bashrc').toString());

$ node index.js
(outputs my bashrc file as expected)
```

**Comment by chase-moskal** [comment]
@nfischer regarding reproducing the bug,

- i first noticed the issue on my archlinux dev machine after a routine system update which took me to node 20.6, then the error popped up when i ran some project build scripts that use shelljs
- later, we found the bug also reproduced in our github actions workflows (on ubuntu-latest)
  - our github actions were set to `node: "20"`
  - we fixed the issue by setting this to `node: "20.5"`
  - our fix was [in this commit](https://github.com/benevolent-games/midnight-grillin/commit/1577ea3daa3d8d90515964b6dd5ae7e2f8f28824) (that fix was successful for this issue, though the build still failed for other totally unrelated issues)

the log for the reproduction on the github action said
```
/home/runner/work/midnight-grillin/midnight-grillin/node_modules/shelljs/src/cat.js:4
common.register('cat', _cat, {
       ^

TypeError: common.register is not a function
```

so perhaps a reproduction would need to specifically involve the `cat` command in shelljs :shrug:

**Comment by chase-moskal** [comment]
@nfischer 

- my project [`@benev/turtle`](https://github.com/benevolent-games/turtle/tree/dev) reproduces this issue
  - on the `dev` branch, commit [77683e8f22d59f93b5283b533663cbebdf46e703](https://github.com/benevolent-games/turtle/commit/77683e8f22d59f93b5283b533663cbebdf46e703)
  - under node 20.6, running `npm i` then `npm run build` should demonstrate this issue
  - see [my results here](https://gist.github.com/chase-moskal/2b53cabeb52014b1e53138d7a374bd40) under node 20.6
  - the build goes fine when i `nvm use 20.5`

my codebase there only makes three function calls with shelljs:
```ts
shell.mkdir("-p", dirname(path))
shell.mkdir("-p", dirname(target.relative))
shell.cp(source.relative, target.relative)
```

however, some quick console logging seems to reveal that the bug kills the program before it gets to these particular lines (i think the bug manifests immediately upon importing shelljs)

**Comment by chase-moskal** [comment]
@nfischer hey maybe this is a commonjs/esm thing?

- the code you posted uses commonjs `var shell = require('shelljs')`
- [my code](https://github.com/benevolent-games/turtle/blob/77683e8f22d59f93b5283b533663cbebdf46e703/s/utils/file_io.ts#L2) and also [this code from remix](https://github.com/remix-run/remix/blob/754248c3072ec88c15d39330127003b4c9112253/integration/deno-compiler-test.ts#L4) are both using `import shell from "shelljs"`

**Comment by webpro** [comment]
> @nfischer hey maybe this is a commonjs/esm thing?
> 
> * the code you posted uses commonjs `var shell = require('shelljs')`
> * [my code](https://github.com/benevolent-games/turtle/blob/77683e8f22d59f93b5283b533663cbebdf46e703/s/utils/file_io.ts#L2) and also [this code from remix](https://github.com/remix-run/remix/blob/754248c3072ec88c15d39330127003b4c9112253/integration/deno-compiler-test.ts#L4) are both using `import shell from "shelljs"`

For the record, [release-it uses ESM/import only](https://github.com/search?q=repo%3Arelease-it%2Frelease-it%20shelljs&type=code) with `"type": "module"` and no compilation whatsoever and has the issue as well starting in Node.js v20.6.0.

**Comment by nfischer** [comment]
@chase-moskal yup that was the trick. This is specific to `.mjs` modules. If I rename my JS file **and** change the import, then the bug reproduces:

```shell
$ cat index.mjs
import shell from "shelljs";

console.log(shell.head('~/.bashrc').toString());

$ node index.mjs
/home/nfischer/shelljs-common-register-v20/node_modules/shelljs/src/cat.js:4
common.register('cat', _cat, {
       ^

TypeError: common.register is not a function
    at Object.<anonymous> (/home/nfischer/shelljs-common-register-v20/node_modules/shelljs/src/cat.js:4:8)
    at Module._compile (node:internal/modules/cjs/loader:1241:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1295:10)
    at Module.load (node:internal/modules/cjs/loader:1091:32)
    at Module._load (node:internal/modules/cjs/loader:938:12)
    at Module.require (node:internal/modules/cjs/loader:1115:19)
    at require (node:internal/modules/helpers:130:18)
    at /home/nfischer/shelljs-common-register-v20/node_modules/shelljs/shell.js:25:3
    at Array.forEach (<anonymous>)
    at Object.<anonymous> (/home/nfischer/shelljs-common-register-v20/node_modules/shelljs/shell.js:24:23)

Node.js v20.6.0
```

**Comment by Pelv** [comment]
Any solution at this problem?

**Comment by brophdawg11** [comment]
I put together a small reproduction of the issue here, I don't know if it's a node issue or a circular dependency issue though: https://github.com/brophdawg11/shelljs-node-20-6-ESM-issue

**Comment by webpro** [comment]
The issue and the fix might be in Node.js: https://github.com/nodejs/node/pull/49500. Looks unreleased yet.

**Comment by The-Code-Monkey** [comment]
@webpro nice find if your keeping track of this, would you be able to ping me here when it is fixed thanks

**Comment by bkuri** [comment]
confirmed fixed as of `v20.6.1` :+1:

**Comment by nfischer** [comment]
Thanks for chasing this down! I'm glad to hear the folks working on Node were able to fix this quickly and cut a new release.
